File: ruby/lib/thelawin/invoice.rb
# frozen_string_literal: true

require "base64"
require "date"

module Thelawin
  # Result object returned after successful invoice generation
  class InvoiceSuccess
    attr_reader :pdf_base64, :filename, :format, :account

    def initialize(pdf_base64:, filename:, format:, account: nil)
      @pdf_base64 = pdf_base64
      @filename = filename
      @format = format
      @account = account
    end

    # @return [Boolean] Always true
    def success?
      true
    end

    # Save the output to a file (PDF or XML depending on format)
    # @param file_path [String] Path to save the file
    def save(file_path)
      dir = File.dirname(file_path)
      FileUtils.mkdir_p(dir) unless File.directory?(dir)
      File.binwrite(file_path, to_bytes)
    end

    # Alias for backwards compatibility
    alias save_pdf save

    # Get the output as bytes
    # @return [String] Binary data (PDF or XML)
    def to_bytes
      Base64.decode64(@pdf_base64)
    end

    # Get the output as a data URL
    # @return [String] Data URL
    def to_data_url
      mime_type = @format.xml_only? ? "application/xml" : "application/pdf"
      "data:#{mime_type};base64,#{@pdf_base64}"
    end

    # Check if the output is XML-only (no visual PDF)
    def xml_only?
      @format.xml_only?
    end

    # Get legal warnings from format detection
    # @return [Array<LegalWarning>]
    def warnings
      @format.warnings
    end
  end

  # Result object returned when invoice generation fails
  class InvoiceFailure
    attr_reader :errors

    def initialize(errors:)
      @errors = errors
    end

    # @return [Boolean] Always false
    def success?
      false
    end
  end

  # Fluent builder for creating invoices
  class InvoiceBuilder
    def initialize(client)
      @client = client
      @format = "auto"
      @profile = "en16931"
      @number = nil
      @date = nil
      @due_date = nil
      @seller = nil
      @buyer = nil
      @items = []
      @payment = nil
      @currency = "EUR"
      @notes = nil
      @leitweg_id = nil
      @buyer_reference = nil
      @tipo_documento = nil
      @template = "minimal"
      @locale = "en"
      @customization = Customization.new
    end

    # Set the output format
    # @param value [String] "auto", "zugferd", "facturx", "xrechnung", "pdf", "ubl", "cii", "peppol", "fatturapa"
    # @return [self]
    def format(value)
      @format = value
      self
    end

    # Set the profile (for ZUGFeRD/Factur-X formats)
    # @param value [String] "minimum", "basic_wl", "basic", "en16931", "extended"
    # @return [self]
    def profile(value)
      @profile = value
      self
    end

    # Set the invoice number
    # @param value [String]
    # @return [self]
    def number(value)
      @number = value
      self
    end

    # Set the invoice date
    # @param value [String, Date] ISO format YYYY-MM-DD or Date object
    # @return [self]
    def date(value)
      @date = value.is_a?(Date) ? value.iso8601 : value
      self
    end

    # Set the due date
    # @param value [String, Date] ISO format YYYY-MM-DD or Date object
    # @return [self]
    def due_date(value)
      @due_date = value.is_a?(Date) ? value.iso8601 : value
      self
    end

    # Set the seller information
    # @param name_or_party [String, Party] Name or Party object
    # @param opts [Hash] Optional attributes (when name is a String)
    # @return [self]
    def seller(name_or_party = nil, **opts)
      if name_or_party.is_a?(Party)
        @seller = name_or_party
      elsif name_or_party.nil? && opts.any?
        @seller = Party.new(**opts)
      else
        @seller = Party.new(name: name_or_party, **opts)
      end
      self
    end

    # Set the buyer information
    # @param name_or_party [String, Party] Name or Party object
    # @param opts [Hash] Optional attributes (when name is a String)
    # @return [self]
    def buyer(name_or_party = nil, **opts)
      if name_or_party.is_a?(Party)
        @buyer = name_or_party
      elsif name_or_party.nil? && opts.any?
        @buyer = Party.new(**opts)
      else
        @buyer = Party.new(name: name_or_party, **opts)
      end
      self
    end

    # Add a line item
    # @param description [String, LineItem] Description or LineItem object
    # @param quantity [Numeric]
    # @param unit_price [Numeric]
    # @param opts [Hash] Optional attributes (unit, vat_rate, natura)
    # @return [self]
    def add_item(description = nil, quantity: nil, unit_price: nil, **opts)
      if description.is_a?(LineItem)
        @items << description
      elsif description.nil? && opts.any?
        @items << LineItem.new(**opts)
      else
        @items << LineItem.new(description: description, quantity: quantity, unit_price: unit_price, **opts)
      end
      self
    end

    # Set multiple line items at once
    # @param items [Array<LineItem, Hash>]
    # @return [self]
    def items(items)
      @items = items.map do |item|
        item.is_a?(LineItem) ? item : LineItem.new(**item)
      end
      self
    end

    # Set payment information
    # @param opts [Hash] Payment attributes
    # @return [self]
    def payment(**opts)
      @payment = opts.is_a?(PaymentInfo) ? opts : PaymentInfo.new(**opts)
      self
    end

    # Set the currency
    # @param value [String] Currency code (default: EUR)
    # @return [self]
    def currency(value)
      @currency = value
      self
    end

    # Set invoice notes/comments
    # @param value [String]
    # @return [self]
    def notes(value)
      @notes = value
      self
    end

    # Set Leitweg-ID for XRechnung (German B2G)
    # @param value [String] e.g., "04011000-12345-67"
    # @return [self]
    def leitweg_id(value)
      @leitweg_id = value
      self
    end

    # Set buyer reference for Peppol
    # @param value [String] Purchase order reference
    # @return [self]
    def buyer_reference(value)
      @buyer_reference = value
      self
    end

    # Set document type for FatturaPA
    # @param value [String] "TD01" (invoice), "TD04" (credit note), etc.
    # @return [self]
    def tipo_documento(value)
      @tipo_documento = value
      self
    end

    # Set the template style
    # @param value [String] "minimal", "classic", or "compact"
    # @return [self]
    def template(value)
      @template = value
      self
    end

    # Set the locale
    # @param value [String] "de", "en", "fr", "es", "it"
    # @return [self]
    def locale(value)
      @locale = value
      self
    end

    # Set logo from file
    # @param file_path [String] Path to logo file
    # @param width_mm [Integer, nil] Logo width in mm
    # @return [self]
    def logo_file(file_path, width_mm: nil)
      content = File.binread(file_path)
      @customization.logo_base64 = Base64.strict_encode64(content)
      @customization.logo_width_mm = width_mm if width_mm
      self
    end

    # Set logo from Base64 string
    # @param base64 [String] Base64 encoded logo
    # @param width_mm [Integer, nil] Logo width in mm
    # @return [self]
    def logo_base64(base64, width_mm: nil)
      @customization.logo_base64 = base64
      @customization.logo_width_mm = width_mm if width_mm
      self
    end

    # Set footer text
    # @param text [String]
    # @return [self]
    def footer_text(text)
      @customization.footer_text = text
      self
    end

    # Set accent color
    # @param color [String] Hex color code
    # @return [self]
    def accent_color(color)
      @customization.accent_color = color
      self
    end

    # Generate the invoice
    # @return [InvoiceSuccess, InvoiceFailure]
    def generate
      errors = validate_required_fields
      return InvoiceFailure.new(errors: errors) unless errors.empty?

      request = build_request
      @client.send(:generate_invoice_internal, request)
    end

    # Validate without generating (dry-run)
    # @return [DryRunResult]
    def validate
      errors = validate_required_fields
      return InvoiceFailure.new(errors: errors) unless errors.empty?

      request = build_request
      @client.send(:validate_invoice_internal, request)
    end

    private

    def validate_required_fields
      errors = []
      errors << { path: "$.invoice.number", code: "REQUIRED", message: "Invoice number is required" } unless @number
      errors << { path: "$.invoice.date", code: "REQUIRED", message: "Invoice date is required" } unless @date
      errors << { path: "$.invoice.seller", code: "REQUIRED", message: "Seller information is required" } unless @seller
      errors << { path: "$.invoice.buyer", code: "REQUIRED", message: "Buyer information is required" } unless @buyer
      errors << { path: "$.invoice.items", code: "REQUIRED", message: "At least one line item is required" } if @items.empty?
      errors
    end

    def build_request
      invoice_data = {
        number: @number,
        date: @date,
        dueDate: @due_date,
        seller: @seller.to_h,
        buyer: @buyer.to_h,
        items: @items.map(&:to_h),
        payment: @payment&.to_h,
        currency: @currency,
        notes: @notes,
        leitwegId: @leitweg_id,
        buyerReference: @buyer_reference,
        tipoDocumento: @tipo_documento
      }.compact

      {
        format: @format,
        profile: @profile,
        template: @template,
        locale: @locale,
        invoice: invoice_data,
        customization: @customization.empty? ? nil : @customization.to_h
      }.compact
    end
  end
end

File: python/thelawin/invoice.py
from __future__ import annotations

import base64
from dataclasses import dataclass, field
from datetime import date
from pathlib import Path
from typing import TYPE_CHECKING, Literal, Union

from .types import (
    Customization,
    GenerateRequest,
    InvoiceData,
    InvoiceFormat,
    InvoiceProfile,
    InvoiceTemplate,
    LineItem,
    Party,
    PaymentInfo,
    ValidationError,
    FormatInfo,
    AccountInfo,
)

if TYPE_CHECKING:
    from .client import ThelawinClient


@dataclass
class InvoiceSuccess:
    pdf_base64: str
    filename: str
    format: FormatInfo
    account: AccountInfo | None = None

    @property
    def success(self) -> Literal[True]:
        return True

    def save_pdf(self, file_path: str | Path) -> None:
        path = Path(file_path)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_bytes(base64.b64decode(self.pdf_base64))

    def to_bytes(self) -> bytes:
        return base64.b64decode(self.pdf_base64)

    def to_data_url(self) -> str:
        return f"data:application/pdf;base64,{self.pdf_base64}"


@dataclass
class InvoiceFailure:
    errors: list[ValidationError]

    @property
    def success(self) -> Literal[False]:
        return False


InvoiceResult = Union[InvoiceSuccess, InvoiceFailure]


@dataclass
class InvoiceBuilder:
    _client: "ThelawinClient"
    _number: str | None = None
    _date: str | None = None
    _due_date: str | None = None
    _seller: Party | None = None
    _buyer: Party | None = None
    _items: list[LineItem] = field(default_factory=list)
    _payment: PaymentInfo | None = None
    _currency: str = "EUR"
    _notes: str | None = None
    _format: InvoiceFormat = "auto"
    _profile: InvoiceProfile = "en16931"
    _template: InvoiceTemplate = "minimal"
    _locale: str = "en"
    _leitweg_id: str | None = None
    _buyer_reference: str | None = None
    _tipo_documento: str | None = None
    _customization: Customization = field(default_factory=Customization)

    def number(self, value: str) -> "InvoiceBuilder":
        self._number = value
        return self

    def date(self, value: str | date) -> "InvoiceBuilder":
        self._date = value.isoformat() if isinstance(value, date) else value
        return self

    def due_date(self, value: str | date) -> "InvoiceBuilder":
        self._due_date = value.isoformat() if isinstance(value, date) else value
        return self

    def seller(
        self,
        name: str,
        *,
        street: str | None = None,
        city: str | None = None,
        postal_code: str | None = None,
        country: str | None = None,
        vat_id: str | None = None,
        email: str | None = None,
        phone: str | None = None,
        peppol_id: str | None = None,
        codice_fiscale: str | None = None,
    ) -> "InvoiceBuilder":
        self._seller = Party(
            name=name, street=street, city=city, postal_code=postal_code,
            country=country, vat_id=vat_id, email=email, phone=phone,
            peppol_id=peppol_id, codice_fiscale=codice_fiscale,
        )
        return self

    def seller_party(self, party: Party) -> "InvoiceBuilder":
        self._seller = party
        return self

    def buyer(
        self,
        name: str,
        *,
        street: str | None = None,
        city: str | None = None,
        postal_code: str | None = None,
        country: str | None = None,
        vat_id: str | None = None,
        email: str | None = None,
        phone: str | None = None,
        peppol_id: str | None = None,
        codice_fiscale: str | None = None,
        codice_destinatario: str | None = None,
        pec: str | None = None,
    ) -> "InvoiceBuilder":
        self._buyer = Party(
            name=name, street=street, city=city, postal_code=postal_code,
            country=country, vat_id=vat_id, email=email, phone=phone,
            peppol_id=peppol_id, codice_fiscale=codice_fiscale,
            codice_destinatario=codice_destinatario, pec=pec,
        )
        return self

    def buyer_party(self, party: Party) -> "InvoiceBuilder":
        self._buyer = party
        return self

    def add_item(
        self,
        description: str,
        quantity: float,
        unit_price: float,
        *,
        unit: str = "C62",
        vat_rate: float = 19.0,
        natura: str | None = None,
    ) -> "InvoiceBuilder":
        self._items.append(
            LineItem(
                description=description, quantity=quantity, unit=unit,
                unit_price=unit_price, vat_rate=vat_rate, natura=natura,
            )
        )
        return self

    def add_item_obj(self, item: LineItem) -> "InvoiceBuilder":
        self._items.append(item)
        return self

    def items(self, items: list[LineItem]) -> "InvoiceBuilder":
        self._items = items
        return self

    def payment(
        self,
        *,
        iban: str | None = None,
        bic: str | None = None,
        terms: str | None = None,
        reference: str | None = None,
    ) -> "InvoiceBuilder":
        self._payment = PaymentInfo(iban=iban, bic=bic, terms=terms, reference=reference)
        return self

    def currency(self, value: str) -> "InvoiceBuilder":
        self._currency = value
        return self

    def notes(self, value: str) -> "InvoiceBuilder":
        self._notes = value
        return self

    def format(self, value: InvoiceFormat) -> "InvoiceBuilder":
        self._format = value
        return self

    def profile(self, value: InvoiceProfile) -> "InvoiceBuilder":
        self._profile = value
        return self

    def template(self, value: InvoiceTemplate) -> "InvoiceBuilder":
        self._template = value
        return self

    def locale(self, value: str) -> "InvoiceBuilder":
        self._locale = value
        return self

    def leitweg_id(self, value: str) -> "InvoiceBuilder":
        self._leitweg_id = value
        return self

    def buyer_reference(self, value: str) -> "InvoiceBuilder":
        self._buyer_reference = value
        return self

    def tipo_documento(self, value: str) -> "InvoiceBuilder":
        self._tipo_documento = value
        return self

    def logo_file(self, file_path: str | Path, width_mm: int | None = None) -> "InvoiceBuilder":
        path = Path(file_path)
        self._customization.logo_base64 = base64.b64encode(path.read_bytes()).decode("utf-8")
        if width_mm is not None:
            self._customization.logo_width_mm = width_mm
        return self

    def logo_base64(self, base64_data: str, width_mm: int | None = None) -> "InvoiceBuilder":
        self._customization.logo_base64 = base64_data
        if width_mm is not None:
            self._customization.logo_width_mm = width_mm
        return self

    def footer_text(self, text: str) -> "InvoiceBuilder":
        self._customization.footer_text = text
        return self

    def accent_color(self, color: str) -> "InvoiceBuilder":
        self._customization.accent_color = color
        return self

    def generate(self) -> InvoiceResult:
        return self._client._generate_sync(self._build_request())

    async def generate_async(self) -> InvoiceResult:
        return await self._client._generate_async(self._build_request())

    def _build_request(self) -> GenerateRequest | InvoiceFailure:
        errors: list[ValidationError] = []

        if not self._number:
            errors.append(ValidationError(path="$.invoice.number", code="REQUIRED", message="Invoice number is required"))
        if not self._date:
            errors.append(ValidationError(path="$.invoice.date", code="REQUIRED", message="Invoice date is required"))
        if not self._seller:
            errors.append(ValidationError(path="$.invoice.seller", code="REQUIRED", message="Seller information is required"))
        if not self._buyer:
            errors.append(ValidationError(path="$.invoice.buyer", code="REQUIRED", message="Buyer information is required"))
        if not self._items:
            errors.append(ValidationError(path="$.invoice.items", code="REQUIRED", message="At least one line item is required"))

        if errors:
            return InvoiceFailure(errors=errors)

        customization = None
        if self._customization.logo_base64 or self._customization.footer_text or self._customization.accent_color:
            customization = self._customization

        return GenerateRequest(
            format=self._format,
            profile=self._profile,
            template=self._template,
            locale=self._locale,
            invoice=InvoiceData(
                number=self._number,  # type: ignore
                date=self._date,  # type: ignore
                due_date=self._due_date,
                seller=self._seller,  # type: ignore
                buyer=self._buyer,  # type: ignore
                items=self._items,
                payment=self._payment,
                currency=self._currency,
                notes=self._notes,
                leitweg_id=self._leitweg_id,
                buyer_reference=self._buyer_reference,
                tipo_documento=self._tipo_documento,
            ),
            customization=customization,
        )
